File: frontend/src/lib/server/storage-paths.ts
import { parse, resolve } from 'node:path';

export function deriveLedgerPath(snapshotPath: string): string {
	const parsed = parse(snapshotPath);
	if (parsed.ext === '') {
		return `${snapshotPath}.sqlite3`;
	}

	return resolve(parsed.dir, `${parsed.name}.sqlite3`);
}

export function deriveJobQueuePath(snapshotPath: string): string {
	const parsed = parse(snapshotPath);
	if (parsed.ext === '') {
		return `${snapshotPath}.jobs.sqlite3`;
	}

	return resolve(parsed.dir, `${parsed.name}.jobs.sqlite3`);
}

export const databasePath =
	process.env.WEATHER_LEDGER_DB_PATH ?? resolve(process.cwd(), '../database/weather.duckdb');

export const ledgerPath = process.env.WEATHER_LEDGER_LEDGER_PATH ?? deriveLedgerPath(databasePath);

export const locationJobQueuePath =
	process.env.WEATHER_LEDGER_JOB_DB_PATH ?? deriveJobQueuePath(databasePath);

File: data-pipeline/storage_paths.py
from __future__ import annotations

import os
from pathlib import Path

DEFAULT_SNAPSHOT_PATH = Path(__file__).resolve().parent.parent / "database" / "weather.duckdb"


def derive_ledger_path(snapshot_path: Path) -> Path:
    if snapshot_path.suffix == "":
        return Path(f"{snapshot_path}.sqlite3")

    return snapshot_path.with_suffix(".sqlite3")


def resolve_snapshot_path() -> Path:
    return Path(os.environ.get("WEATHER_LEDGER_DB_PATH", DEFAULT_SNAPSHOT_PATH))


def resolve_ledger_path() -> Path:
    return Path(
        os.environ.get(
            "WEATHER_LEDGER_LEDGER_PATH",
            derive_ledger_path(resolve_snapshot_path()),
        )
    )
